src/uct/base/uct_iface.h

```diff
@@ -406,38 +406,66 @@ typedef struct uct_iface_local_addr_ns {
     void uct_##_name##_cleanup(void);
 
 
+#define _UCT_IFACE_UNQUOTE(_x) _x
+#define _UCT_IFACE_CTOR_
+#define _UCT_IFACE_DTOR_
+#define _UCT_IFACE_CTOR_ctor UCS_F_CTOR
+#define _UCT_IFACE_DTOR_ctor UCS_F_DTOR
+#define _UCT_IFACE_CTOR_static static
+#define _UCT_IFACE_DTOR_static static
+
 /**
- * Register component and TL
+ * Register/unregister TL
  *
- * @param [in] _name   Component and TL name
+ * @param [in] _name          Component and TL name
+ * @param [in] _scope         Scope for functions, must be ctor, static or empry
+ * @param [in] _init_code     Initialization code
+ * @param [in] _cleanup_code  Cleanup code
  */
-#define UCT_TL_INIT(_name) \
-    void uct_##_name##_init(void); \
-    static void uct_component_tl_##_name##_ctor(void); \
-    void uct_##_name##_init(void) { \
-        uct_component_tl_##_name##_ctor(); \
-        uct_component_##_name##_ctor(); \
+#define UCT_TL_INIT(_name, _scope, _init_code, _cleanup_code) \
+    _UCT_IFACE_UNQUOTE(_UCT_IFACE_CTOR_##_scope) void uct_##_name##_init(void) \
+    { \
+        _init_code; \
         uct_tl_##_name##_ctor(); \
     } \
-    void uct_component_tl_##_name##_ctor(void)
+    _UCT_IFACE_UNQUOTE(_UCT_IFACE_DTOR_##_scope) void uct_##_name##_cleanup(void) \
+    { \
+        uct_tl_##_name##_dtor(); \
+        _cleanup_code; \
+    }
 
 
 /**
- * Unregister component and TL
+ * Register/unregister component and TL
  *
- * @param [in] _name   Component and TL name
+ * @param [in] _name          Component and TL name
+ * @param [in] _scope         Scope for functions, must be ctor, static or empry
+ * @param [in] _init_code     Initialization code
+ * @param [in] _cleanup_code  Cleanup code
  */
-#define UCT_TL_CLEANUP(_name) \
-    void uct_##_name##_cleanup(void); \
-    static void uct_component_tl_##_name##_dtor(void); \
-    void uct_##_name##_cleanup(void) { \
-        uct_tl_##_name##_dtor(); \
-        uct_component_##_name##_dtor(); \
-        uct_component_tl_##_name##_dtor(); \
-    } \
-    void uct_component_tl_##_name##_dtor(void)
+#define UCT_SINGLE_TL_INIT(_name, _scope, _init_code, _cleanup_code) \
+    UCT_TL_INIT(_name, _scope, \
+                {_init_code; uct_component_##_name##_ctor();}, \
```

Comment:
can we call `uct_base_tl_register(_component, &uct_##_name##_tl)`;  here instead of `uct_component_##_name##_ctor(),` and this way remove uct_component_##_name##_ctor altogether

Reply:
so instead of defining component ctor/dtor in UCT_COMPONENT_REGISTER_DEF, can we define helper function like uct_component_register(&component) and call them from UCT_SINGLE_TL_INIT and from IB ctor?
seems like every component is either using UCT_SINGLE_TL_INIT, or defining its own ctor (uct_ib_init)

Reply:
> so instead of defining component ctor/dtor in UCT_COMPONENT_REGISTER_DEF, can we define helper function like uct_component_register(&component) and call them from UCT_SINGLE_TL_INIT and from IB ctor?
> seems like every component is either using UCT_SINGLE_TL_INIT, or defining its own ctor (uct_ib_init)

done